bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/discovery/SerialNumberHelper.java

```diff
@@ -0,0 +1,92 @@
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.client.api.ContentResponse;
+import org.eclipse.jetty.http.HttpStatus;
+import org.openhab.binding.hdpowerview.internal.dto.UserData;
+import org.openhab.binding.hdpowerview.internal.dto.gen3.Info;
+import org.openhab.core.io.net.http.HttpClientFactory;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonSyntaxException;
+
+/**
+ * This helper class fetches the hub serial number so it can be used as the representation property
+ *
+ * @author Andrew Fiddian-Green - Initial contribution
+ */
+@NonNullByDefault
+@Component(service = SerialNumberHelper.class)
+public class SerialNumberHelper {
+
+    public enum ApiVersion {
+        V1,
+        V3;
+    }
+
+    private static final String API_V1_URL_FORMAT = "http://%s/api/userdata";
+    private static final String API_V3_URL_FORMAT = "http://%s/gateway/info";
+
+    private final Gson gsonParser = new Gson();
+    private final HttpClientFactory httpClientFactory;
+
+    @Activate
+    public SerialNumberHelper(@Reference HttpClientFactory httpClientFactory) {
+        this.httpClientFactory = httpClientFactory;
+    }
+
+    /**
+     * Get the serial number of the hub / gateway on the given IPv4 address
+     *
+     * @param ipAddress a dotted ipv4 address
+     * @param apiVersion the hub / gateway api version used
+     * @return the serial number
+     */
+    public String getSerialNumber(String ipAddress, ApiVersion apiVersion) {
+        try {
+            HttpClient httpClient = httpClientFactory.getCommonHttpClient();
+            String uri = String.format(ApiVersion.V1 == apiVersion ? API_V1_URL_FORMAT : API_V3_URL_FORMAT, ipAddress);
+            ContentResponse content = httpClient.GET(uri);
+            if (HttpStatus.OK_200 == content.getStatus()) {
+                String json = content.getContentAsString();
+                switch (apiVersion) {
+                    case V1:
+                        UserData userData = gsonParser.fromJson(json, UserData.class);
+                        if (userData != null && userData.serialNumber instanceof String serialNumber) {
+                            return serialNumber;
+                        }
+                        break;
+
+                    case V3:
+                        Info info = gsonParser.fromJson(json, Info.class);
+                        if (info != null && info.getSerialNumber() instanceof String serialNumber) {
+                            return serialNumber;
+                        }
+                        break;
+                }
+            }
+        } catch (JsonSyntaxException | InterruptedException | ExecutionException | TimeoutException e) {
+        }
+        // fall back to ip address
+        return ipAddress;
+    }
```

Comment:
This method does two different and mutually exclusive things, namely calls either API v1 or API v3. So I'm wondering if it wouldn't be better to simplify by splitting it, i.e.:
```suggestion
    public String getSerialNumberApiVersion1(String ipAddress) {
        try {
            HttpClient httpClient = httpClientFactory.getCommonHttpClient();
            ContentResponse content = httpClient.GET(API_V1_URL_FORMAT);
            if (HttpStatus.OK_200 == content.getStatus()) {
                String json = content.getContentAsString();
                UserData userData = gsonParser.fromJson(json, UserData.class);
                if (userData != null && userData.serialNumber instanceof String serialNumber) {
                    return serialNumber;
                }
            }
        } catch (JsonSyntaxException | InterruptedException | ExecutionException | TimeoutException e) {
        }
        // fall back to ip address
        return ipAddress;
    }
```

and likewise for API version 3?

Reply:
.. but yes we can split it if you prefer..

Reply:
It seems cleaner, so I'd prefer that.

Reply:
Done.